README.md
# FrameWork.php #
### Developed by Shaun Bebbington and Linden Bryon ###
#### Special thanks to Aaron Zvimba for his help, advice and support ####

FrameWork.php started out life as a simple and quite messy OO MVC framework in order to mentor a junior PHP developer whilst I was working at Rushcliff Ltd in Burton-Upon-Trent, Staffordshire. The idea was to link a controller to each page, and have a model per controller, so that the controller class would act as a middle-man between the front-end view and the database as necessary.

I also thought it necessary to allow numerous file extensions on the page uri alias, so for instance, in the skeleton provided, one might use any of the following:

* http://framework.local/home
* http://framework.local/home.php
* http://framework.local/home.aspx
* http://framework.local/home.htm
* http://framework.local/home.html

The idea was to obscure that it was a PHP framework, and to send the page URLs with the .aspx file extension to our pen testers. I don't know if this was a good idea, however no issues that came back requiring attention even after we let them know that it was a PHP framework. One sundry benefit to this approach is that I've accidentially provided a way for websites to migrate to this framework from another tech stack without losing any SEO ratings that they have. For instance, if your categories.aspx page ranked highly, you are still able to use this extension on a site here. I handle this by using canonical tags in the header by default if a file extension is detected and the route is valid.

I never had the chance to finish the framework to a satisfactory degree at Rushcliff Ltd, so although it has been deployed on a live product, it remains imperfect. I was allowed to continue developing it on leaving [Rushcliff] so here it is. Still imperfect but always improving. This is not something for those hipster and artisan developer-types, it is fairly simple and fairly stupid.

Linden has set up on GitHub so I have invited him as a collaborator on this project. Linden has expertise with MySQL, PHP 5/7, HTML 5/CSS 3, JavaScript (jQuery) and some JavaScript frameworks, and is a very keen and highly recommended developer.

If you have any comments then please feel free to contact me.

Shaun Bebbington.
Twitter: @YearOfCodes
<hr />
<img src="https://sonarcloud.io/api/project_badges/measure?project=coffee&metric=sqale_rating" alt="Scanned with SonarCloud" />


CHANGELOG.md
# Project FrameWork.php MVC v1.0.0-RC1 #

This is a fairly simple and fairly stupid MCV framework for PHP 7. Simply set up your path in the allowed segments, the name of your path will point to the /application/view folder - you must place a view file with the same name as the allowed path with a .phtml extension, so in the example home in the $allowedSegments resource in the FrameworkCore.php will load the home.phtml view in /application/view

#### File Structure: ####

	../application
		--> config
			\ This is where to set the database and site credentials and such
		--> controller
			\ For all of the custom controller methods and logic
		--> core
			\ Contains the main engine room for looking up page requests,
				and loading the right controller and view
		--> model
			\
			--> read
			|	\ For read access users to database
			--> write
			|	\ For read/write access to database
			\ main Db connection and model logic
		--> library
			\ Helper functions and other such things
		--> view
			\ HTML (with embedded PHP) views
		--> partial
			\ For partial HTML views such as headers, footers and menus

--


	@version	1.0.0-RC1
	@date		February 2016 - current date
	@author		Shaun Bebbington (version 0.0.1 to current)
				&& Linden Bryon (version 0.0.1 to 0.0.7)
	@changes as of 2016-02-19:
				Added partial views in seperate directory
				Added model and controller core so all other model and
				controller classes extend this as required Controller
				core file will include the library class for all
				helper functions Default to Bootstrap 3.3.x and jQuery
				2.2.x Application folder moved outside of the public
				facing directory Admin area added for future use and
				expansion
				Fixed: Chooses correct segment (may need changing
				according to URL structure)
	@changes as of 2016-05-12:
				Fixed: Session Cookie handling correctly for HTTP and HTTPS
				connections
	@changes as of 2016-06-02:
				Forgot to say that the TimeAndDate class in the library has
				been fixed
	@changes as of 2016-06-09:
				Removed much of the functionality to the Framework Core class
				Added global methods for file routing to the application folder
				You can now relate controllers to views; if a view has a
				related controller, the controller is automatically required
				and an instance of which is initiated in the $this->controller
				object set in the core. Partial views are set in the core
				under $this->partial['partial-instance'], to use this, use
				require_once($this->partial['partial-instance']); in the
				PHP view
	@changes as of 2016-09-12:
				Can now utilise ZF-style view variables if using
				$this->view->objName in the controllers; this becomes
				$this->objName in the PHP/HTML view files, meaning you no
				longer require $this->controller->instance->view->objName
				Note: that anything set in the view object will overwrite
				any other objects of the same name
	@changes as of 2016-09-14:
				Now has specific flash message for session cookie, use
				$this->setFlashMessage($k, $v); in the controllers where $k
				is the key and $v is the value, and to check if the flash
				message has been set, use $this->getFlashMessage($k), which
				will check if the key is set. In the view, use $this->flash->$k
				to get the value.
	@changes as of 2017-01-06:
				Namespaces beginning to be added for more flexibility, renaming
				project and other tidying up to improve the stuff; as of this
				date all updates and improvements will be handled by
				Shaun Bebbington
	@changes as of 2017-01-11:
				Minor refactoring to the methods in the library
	@changes as of 2017-01-12:
				Some more refactoring and an initial commit to GitHub, the main
				thing is to move the global helpers to a file in the
				application/core directory, therefore making this file quite
				light-weight in PHP terms. All of the page handling stuff is
				handled in the Core directly now; Need to add in the HTML
				builder and other things, as well as work on the modelling.
	@changes as of 2017-01-13:
				I've added a convertSnakeCase function to the library so that
				database keys may be used in a PHP-friendly way such as
				camelCase. In my test, sub_header in the database becomes
				subHeader in my PHP view - much easier to remember.
	@changes as of 2017-01-23:
				I've added in the HTML builder with an example of how to
				use on the home page view in application/view/home.phtml.
				This will allow you to generate much of your HTML using PHP
				but will also produce the HTML in minimised form; as PHP
				doesn't care so much about white spaces, you can align your
				PHP in nice way without affecting the outcome of the HTML
				(still generated as minimised regardless of how you format
				your PHP).
				Some of the SEO companies that I have worked with like
				minimised HTML as it might affect the page load time by
				a few milliseconds if you're lucky. I've also started to
				use the strict PHP 7 types in parts of the project, it should
				be easy enough to fix this for PHP 5.x version, but from
				now on I'll be using the latest PHP standards, so consider
				it now a PHP 7 framework.
	@changes as of 2017-02-06:
				Added in a path router to the core - for instance, if you
				have in your view folder /admin/login.phtml and
				/admin/dashboard.phtml, you will now be able to use this path
				as a URI request. Setting up the $allowedSegments and
				$pageControllers has not changed in the Framework Core, as
				it will build the URI string and route to it. I've also
				improved the logic for working out which view to load in the
				loadPage() method of the core; now it will check if the file
				exists as well as a check to see if the URI request is allowed.
	@changes as of 2017-02-08:
				I've been in contact with my former colleague Linden Bryon

application/library/Library.php
<?php
namespace Application\Library;

use Application\Core\FrameworkException\FrameworkException;
if (! defined('FRAMEWORKPHP') || FRAMEWORKPHP != 65535) {
    require_once ("../view/403.phtml");
}

class Library
{

    private $key;

    private $encryption;

    public function __construct()
    {
        $this->key = getConfig('key');
        $this->encryption = getConfig('encryptionType');
    }

    /**
     * For debugging objects, optional die and die message included
     * Parameters now allow for a header for each object
     *
     * @param mixed $variable
     * @param bool $die
     * @param string $message
     * @param string $file
     * @param string $line
     * @param string $header
     * @author sbebbington && Linden
     * @date 26 Sep 2017 09:47:37
     * @version 1.0.0-RC1
     * @return null
     */
    public function debug($variable = null, bool $die = false, string $message = '', string $file = '', string $line = '', string $header = ''): void
    {
        if (getConfig('mode') != 'test') {
            return;
        }
        if (is_null($variable)) {
            echo '<p>You need to send the scalar or resource that you are trying to debug as your first parameter, else this don\'t work</p>';
            exit();
        }
        echo (strlen($header) > 0) ? "<div><h1>{$header}</h1>" : "";
        echo '<pre>' . print_r($variable, 1) . '</pre>';
        echo $file != null ? '<pre>File: ' . print_r($file, 1) . '</pre>' : "";
        echo $line != null ? '<pre>Line: ' . print_r($line, 1) . '</pre>' : "";
        echo (strlen($header) > 0) ? "</div>" : "";
        if (isTrue($die)) {
            die("{$message}");
        }
    }

    /**
     * For seeing contents of variables or objects, optional die
     * and die message included
     *
     * @param object $variable
     * @param bool $die
     * @param string $message
     * @param string $file
     * @param string $line
     * @author sbebbington && Linden
     * @date 26 Sep 2017 09:47:15
     * @version 1.0.0-RC1
     * @return null
     */
    public function dump($variable, bool $die = false, string $message = '', string $file = '', string $line = ''): void
    {
        if (getConfig('mode') != 'test') {
            return;
        }
        var_dump($variable);
        echo $file != null ? '<pre>File: ' . print_r($file, 1) . '</pre>' : "";
        echo $line != null ? '<pre>Line: ' . print_r($line, 1) . '</pre>' : "";
        if (isTrue($die)) {
            die($message);
        }
    }

    /**
     * Returns current version of the framework
     *
     * @author sbebbington B
     * @date 26 Sep 2017 10:05:31
     * @return string
     */
    public function version(): string
    {
        return '1.0.0-RC1';
    }

    /**
     * Is it Easter yet?
     * Try <?= $this->controllerInstance->libraryInstance->easterEgg(); ?> in your view
     *
     * @author Shaun B
     * @date 30 Aug 2018 16:29:57
     * @return "Something good"
     */
    public function easterEgg(string $id = "something-good"): string
    {
        $easterEgg = chr(116) . chr(104) . chr(101) . chr(99) . chr(97) . chr(116) . chr(97) . chr(112) . chr(105);
        return trim("
            <div id=\"{$id}\" class=\"container\">
                <a href=\"http://{$easterEgg}.com\">
                    <img src=\"http://{$easterEgg}.com/api/images/get?format=src&type=gif\" alt=\"Easter Egg\" />
                </a>
            </div>
        ") ?? "Something good";
    }

    /**
     * Password encryption generator
     *
     * @param string $string
     * @param string $secret
     * @param int $padding
     * @param bool $urlEncode
     * @author sbebbington && Stack Overflow
     * @date 1 Mar 2017 08:54:14
     * @version 1.0.0-RC1
     * @return string
     */
    public function encryptIt(string $string, string $secret = '', int $padding = 8, bool $urlEncode = false): string
    {
        $md5 = ($secret === '') ? md5(md5($this->key)) : md5(md5($secret));
        $encrypt = $this->getEncryptionPadding($padding) . openssl_encrypt($string, $this->encryption, $md5) . $this->getEncryptionPadding($padding);
        return (isTrue($urlEncode)) ? urlencode($encrypt) : "{$encrypt}";
    }

    /**
     * Password decryption generator
     *
     * @param string $string
     * @param string $secret
     * @param int $padding
     * @param bool $urlDecode
     * @author sbebbington && Stack Overflow
     * @date 1 Mar 2017 08:57:23
     * @version 1.0.0-RC1
     * @return string
     */
    public function decryptIt(string $string, string $secret = '', int $padding = 8, bool $urlDecode = false): string
    {
        $md5 = ($secret === '') ? md5(md5($this->key)) : md5(md5($secret));
        $decrypt = openssl_decrypt(substr($string, $padding, - $padding), $this->encryption, $md5);
        return (isTrue($urlDecode)) ? urldecode($decrypt) : "{$decrypt}";
    }

    /**
     * Redirects using the PHP header command
     *
     * @param string $destination
     * @param string $host
     * @param int $serverResponseCode
     * @author sbebbington || Steve
     * @date	17 Nov 2017 10:46:39
     * @version 1.0.0-RC1
     * @return void
     */
    public function redirect(string $destination = '', string $host = '', int $serverResponseCode = 307): void
    {
        if ($destination == '' || $host == '') {
            $this->debug("You need to set a destination and host parameters as a string to call the Library redirect() method", true);
        }
        $host = preg_replace('/^https?\:\/\//', '', $host);
        $http = isHttps() ? "https://" : "http://";
        header("Location:{$http}{$host}/{$destination}", true, $serverResponseCode);
        exit();
    }

    /**
     * Will add a random and predictable padding
     * to the encrypted and decrypted string.
     * I Made this method less like a ZX80 sub
     * routine (I had been experimenting with
     * ZX80 BASIC on that day so appologies)
     *
     * @param int $numberToPad
     * @author sbebbington
     * @date 1 Mar 2017 09:01:05
     * @version 1.0.0-RC1
     * @return string
     */
    public function getEncryptionPadding(int $numberToPad = 8): string
    {
        $shuffle = str_shuffle("1q2w3e4r5t6y7u8i9o0p!AS£D\$%F^G!H*J(K)L-z=x[c]v{b}n;m:QW@E#R*T<Y>U,I.O/P?a|s%d1f2g3h4j5k6l7Z8X9C0VBNM");
        
        return substr($shuffle, 0, $numberToPad);
    }

    /**
     * Redirects by using HTML/JavaScript
     *
     * @param string $destination
     * @param string $website
     * @author sbebbington
     * @date 2 Feb 2017 13:18:50
     * @version 1.0.0-RC1
     * @return void
     * @throws FrameworkException
     */
    public function redirectExternal(string $destination = '', string $website = null): void
    {
        if ($destination == '' || $website == '') {
            throw new FrameworkException("You need to set a destination and host parameters as a string to call the Library redirectExternal() method");
        }
        die(trim("
            <!DOCTYPE html>
            <html lang=\"en-gb\">
                <head>
                    <meta charset=\"UTF-8\">
                    <title>Please wait, redirecting to {$website}</title>
                    <script type=\"text/javascript\">window.location.href=\"{$destination}\"</script>
                </head>
                <body></body>
            </html>
        "));
    }

    /**
     * Unit testing for any method with up to 10 parameters
     * Send the parent object, internal method name, any
     * parameters as a flat array and the expected result
     * to use this
     *
     * @param
     *            object, string, array, any, [boolean]
     * @author sbebbington
     * @date	21 Feb 2018 09:53:55
     * @version 1.0.0-RC1
     * @return boolean | string
     */
    public function testUnit($object = null, string $method = null, $params = array(), $expectedResult = null, bool $tested = false)
    {
        if ($this->checkUnitTestParameters($object, $method, $expectedResult)) {
            return false;
        }
        $passCol = "color: green;";
        $failCol = "color: red;";
        
        if (! is_array($params) && (is_string($params) || is_numeric($params))) {
            $pass = $object->{$method}($params);
            $tested = true;
        } else if (is_array($params)) {
            $pass = $object->{$method}($params[0], $params[1] ?? null, $params[2] ?? null, $params[3] ?? null, $params[4] ?? null, $params[5] ?? null, $params[6] ?? null, $params[7] ?? null, $params[8] ?? null, $params[9] ?? null);
            
            $tested = true;
        }
        
        return ($tested === true) ? $this->outputUnitTestResult($passCol, $failCol, $expectedResult, $pass) : print("<p>Please send the parameters as an array, a string or a numeric value</p>");
    }

    /**
     * Does a check for the main parameters sent to
     * $this->testUnit()
     *
     * @param mixed $object
     * @param string $method
     * @param mixed $expectedResult
     * @author Shaun B
     * @date	6 Aug 2018 13:21:07
     * @throws FrameworkException
     */
    public function checkUnitTestParameters($object, $method, $expectedResult): bool
    {
        if (getConfig('mode') != 'test') {
            return false;
        }
        if ($object === null || ! is_object($object)) {
            throw new FrameworkException("Pass an object to this method in order to test it");
        }
        if ($method === null) {
            throw new FrameworkException("You need to specify the name of the method that you want to test");
        }
        if ($expectedResult == null) {
            throw new FrameworkException("You need to specify your expected return value from the method that you are testing");
        }
        return true;
    }

    /**
     * Shows the test results in <P> tags
     *
     * @param string $passCol
     * @param string $failCol
     * @author Shaun B
     * @date	6 Aug 2018 13:25:49
     */
    public function outputUnitTestResult(string $passCol, string $failCol, $expectedResult = null, $pass = null): bool
    {
        if (isTrue($tested)) {
            echo "<p style=\"";
            echo ($pass == $expectedResult) ? "{$passCol}\">Test matched expected result" : "{$failCol}\">Test failed";
            echo "</p>" . PHP_EOL;
            echo "Expected: {$expectedResult}" . PHP_EOL;
            echo "Actual: {$pass}" . PHP_EOL;
            return ($pass == $expectedResult);
        }
        return false;
    }

    /**
     * This will convert the snake_case stuff typically used in databases
     * to normal camelCase typically used in PHP
     *
     * @param
     *            string
     * @author sbebbington
     * @date 6 Jul 2017 12:14:42
     * @version 1.0.0-RC1
     * @return string
     */
    public function convertSnakeCase(string $snake = null, string $delimiter = '_')
    {
        if ($snake !== null) {
            $_snake = explode($delimiter, $snake);
            if (count($_snake) == 1) {
                return $snake;
            }
            $camelBuilder = '';
            foreach ($_snake as $key => $data) {
                $camelBuilder .= ($key === 0) ? $data : ucfirst($data);
            }
            return $camelBuilder;
        }
        return "{$snake}";
    }

    /**
     * Converts to camelCase where one or more dashes
     * appear in the string
     *
     * @param
     *            string
     * @author sbebbington
     * @date 6 Jul 2017 12:17:34
     * @version 1.0.0-RC1
     * @return string
     */
    public function camelCaseFromDashes($string)
    {
        return $this->convertSnakeCase($string, '-');
    }

    /**
     * This will convert a camelCase string to
     * snake_case as database enthusiasts like
     * snake_case.
     * A lot
     *
     * @param
     *            string, int
     * @author sbebbington
     * @date 3 Feb 2017 13:46:37
     * @version 1.0.0-RC1
     * @return string
     */
    public function convertToSnakeCase(string $unSnaked = null, int $offset = 0)
    {
        if ($unSnaked === null || $unSnaked === '') {
            return '';
        }
        $index = $charBuffer = 0;
        $stringBuffer = '';
        while ($index < strlen($unSnaked)) {
            $charBuffer = ord($unSnaked[$index]);
            if ($index > $offset) {
                if ($charBuffer < 91 && $charBuffer > 64) {
                    $charBuffer += 32;
                    $stringBuffer .= '_';
                }
            }
            $stringBuffer .= chr($charBuffer);
            $index ++;
        }
        return "{$stringBuffer}";
    }

    /**
     * Cleanses and trims the data, used for posted data etc...
     * The recursion bit has been removed from the original
     * version as it seemed a little over-kill, although quite
     * clever; it will also handle HTML Special Chars if necessary
     *
     * @param
     *            array, boolean, [array]
     * @author sbebbington && Vietnam
     * @date 6 Jan 2017 15:36:29
     * @version 1.0.0-RC1
     * @return array
     */
    public function cleanseInputs($data, bool $htmlSpecialChars = false, $cleanInput = array())
    {
        foreach ($data as $key => $value) {
            $cleanInput[$key] = (isFalse($htmlSpecialChars)) ? trim(strip_tags($value)) : htmlspecialchars(trim($value));
        }
        return $cleanInput;
    }

    /**
     * Used for identifying .local and .localhost sites to set
     * PHP error reporting - send true to the method to return
     * the subdomain, i.e., staging or test, or whatever the
     * Reapit is for staging domains
     *
     * @param
     *            boolean
     * @author sbebbington
     * @date 10 Jan 2017 09:25:07
     * @version 1.0.0-RC1
     * @return string
     */
    public function domainType(bool $subDomain = false)
    {
        $host = explode(".", $this->host());
        return (isFalse($subDomain)) ? $host[count($host) - 1] : $host[0];
    }

    /**
     * This will perform a soft minimisation of
     * your javascript [jQuery] files; As it is
     * a dumb minimisation there are some a few
     * limitations, for instance:
     * [1] All comments must be /* style
     * [2] Variable names are not obfustated
     * [3] Not all white spaces are removed
     * [4] It will not remove the final ;
     * where it is not necessary
     *
     * @param
     *            string, [boolean], [boolean]
     * @author sbebbington
     * @date 21 Feb 2017 15:20:10
     * @version 1.0.0-RC1
     * @return string
     */
    public function softMinimiseJS(string $filePathName = null, bool $doubleSpaces = true, bool $spacedTab = true)
    {
        if (empty($filePathName)) {
            return '';
        }
        $remove = array(
            'remove' => array(
                "\t",
                "\r\n"
            ),
            'replace' => array(
                '+' => " + ",
                '==' => " == ",
                '!=' => " != ",
                'if(' => "if (",
                '++' => "++ ",
                '===' => " === ",
                '!==' => " !== ",
                '{' => " { ",
                'space' => "  ",
                'tab' => "    "
            )
        );
        $file = file_get_contents($filePathName);
        foreach ($remove['remove'] as $data) {
            $file = str_replace($data, '', $file);
        }
        foreach ($remove['replace'] as $_key => $data) {
            if ($_key === 'space' || $_key === 'tab') {
                if ($_key == 'space' && isFalse($doubleSpaces)) {
                    continue;
                }
                if ($_key === 'tab' && isFalse($spacedTab)) {
                    continue;
                }
                $_key = '';
            }
            $file = str_replace($data, $_key, $file);
        }
        return $file;
    }

    /**
     * Handles the posting of data
     *
     * @param
     *            string, object, string, string, string
     * @author sbebbington && Stack Overflow
     * @date 3 Mar 2017 09:51:09
     * @version 1.0.0-RC1
     * @return string
     */
    public function filePostContents(string $url, $data, string $applicationType = 'x-www-form-urlencoded', string $username = '', string $password = '', string $characterEncoding = 'utf-8'): string
    {
        if ($applicationType === 'x-www-form-urlencoded') {
            if (! is_object($data) || ! is_array($data)) {
                $data = [
                    $data
                ];
            }
            $data = http_build_query($data);
        }
        $options = array();
        
        $options['http'] = array(
            'method' => 'POST',
            'content' => $data
        );
        $header = array(
            'header' => "Content-type: application/{$applicationType};charset={$characterEncoding}"
        );
        if (is_string($data)) {
            $header['header'] .= PHP_EOL . 'Content-Length: ' . strlen($data) . PHP_EOL;
        }
        if (! empty($username) && ! empty($password)) {
            $header['header'] .= PHP_EOL . "Authorization: Basic " . base64_encode("{$username}:{$password}");
        }
        $options['http'] = array_merge($options['http'], $header);
        
        $context = stream_context_create($options);
        return file_get_contents($url, false, $context);
    }
}
